drivers: stm32: SPI: cannot send several buffers if frame size=16bits

dgastonochoa opened the pull request:
The stm32 SPI driver does not work if one tries to send several buffers with a frame size of 16 bits.

If the SPI's transaction's frame size is set to 16 bits and more than one buffer wants to be sent, the data sent is wrong. This is caused by a function dividing the length of the 'next buffer' by a parameter `dfs`, which should not happen.

More details in #57247 

The following proposal assumes that the `struct spi_buf`'s `len` field is expected to be units of frame size, not bytes. This, as far as I understand, matches the logic of `spi_context_update_tx`.

After removing the division by `dfs`, the result is:

![after-fix](https://user-images.githubusercontent.com/127399215/234312352-0d7e65c7-e3a2-4827-a03e-6697deda9f24.png)

Fixes #57247 
Closes #57247

dgastonochoa committed:
drivers: stm32: SPI: cannot send several buffers if frame size is 16 bits

spi_context_get_next_buf must not divide `len` by `dfs` because, in SPI,
buffer lengths are given in units of data (in this case, 16 bits), not in
bytes.

Signed-off-by: Daniel Gaston Ochoa <dgastonochoa@gmail.com>

anangl reviewed (changes requested):
> The following proposal assumes that the `struct spi_buf`'s `len` field is expected to be units of frame size, not bytes. This, as far as I understand, matches the logic of `spi_context_update_tx`.


> The following proposal assumes that the `struct spi_buf`'s `len` field is expected to be units of frame size, not bytes. This, as far as I understand, matches the logic of `spi_context_update_tx`.

This is a wrong assumption. `spi_context_update_tx()` updates `ctx->tx_len` with `len`, not `dfs * len`, because `ctx->tx_len` is initialized in `spi_context_get_next_buf()` (called initially from `spi_context_buffers_setup()`) with the buffer length divided by `dfs`.

See also https://github.com/zephyrproject-rtos/zephyr/issues/57247#issuecomment-1523507842.

dgastonochoa committed:
drivers: stm32: SPI: cannot send several buffers if frame size is 16 bits

First `spi_context_buffers_setup` must use a `dfs` of 1 or 2 depending on
the frame size.

Signed-off-by: Daniel Gaston Ochoa <dgastonochoa@gmail.com>

carlescufi commented:
@erwango @FRASTM @gautierg-st @ABOSTM could you please take a look?
